### DISCUSSION TITLE
SunbirdEd-mobile-app: Import Content functionality not working.

### DISCUSSION THREAD

**Discussion by chandu4002** [Sunbird-Ed/Community#703] [Bugs]
Import content functionality from side-menu is not working, while selecting .ecar file from local storage it comes back to app and nothing happens. 

**Steps to replicate:**
1.Open sunbird mobile app, click on side-menu.
2.Then click on Import Content Option.
3.find .ecar file in device local storage and select it.
4.then It comes back to app and nothing happens.

**Note: FYI, After selecting .ecar file when it comes back to app the following code is getting executed. In the below code resolve(actionsTobeDone) giving empty array "[]"**

        async handleSunbirdSplashScreenActions(): Promise<undefined> {
                const stringifiedActions = await new Promise<string>((resolve) => {
                    if(this.platform.is('android') && splashscreen){
                        splashscreen.getActions((actionsTobeDone) => {
                            **resolve(actionsTobeDone);**
                        });
                    }
                });
        const actions: { type: string, payload: any }[] = JSON.parse(stringifiedActions);

        if (actions.length) {
            for (const action of actions) {
                switch (action.type) {
                    case 'TELEMETRY': {
                        await this.splashScreenTelemetryActionHandlerDelegate.onAction(action.payload).toPromise();
                        break;
                    }
                    case 'IMPORT': {
                        await this.splashScreenImportActionHandlerDelegate.onAction(action.payload).toPromise();
                        break;
                    }
                    case 'DEEPLINK': {
                        await this.splashScreenDeeplinkActionHandlerDelegate.onAction(action.payload).toPromise();
                        break;
                    }
                    default:
                        return;
                }
            }
        }
        if(splashscreen){
            splashscreen.markImportDone();
            splashscreen.hide();
        }
    } 

cc: @swayangjit

**Comment by simranKtrigyn** [comment]
@swayangjit the filepath we are getting for import is: file:///storage/emulated/0/Download/prshnottrii_1700289400353_do_3139280384301137921472_2 1.ecar'

With this we are getting IMPORT FAILED Error

<img width="323" alt="Screenshot 2024-03-22 at 8 44 37 PM" src="https://github.com/Sunbird-Ed/Community/assets/129143253/e75dd7b3-23c9-4108-973a-b951412f4df3">
<img width="321" alt="Screenshot 2024-03-22 at 8 45 06 PM" src="https://github.com/Sunbird-Ed/Community/assets/129143253/31f8b736-b00e-4440-ab46-bfa0608b4b7f">


Could you please check the file path if it is in correct format. Please Help

**Reply by swayangjit** [reply]
@simranKtrigyn  @chandu4002  [Similar type of issue]([url](https://project-sunbird.atlassian.net/browse/ED-3677)) our QA also reported . We are working on the solution. You can try checking this devices less than 12. I suspect this is some permission issue.

**Reply by chandu4002** [reply]
@swayangjit Thanks for the response, we have tried it in Android 11 device still getting the same error(**Import Failed**).
FYI, from **importEcar()** method we are getting following response.
<img width="194" alt="Screenshot 2024-03-26 at 9 50 18 AM" src="https://github.com/Sunbird-Ed/Community/assets/129141695/5012b2b3-1998-4998-b1f1-6dfe8929a676">

**following is the importEcar() method code for your reference**


        this.contentService.importEcar({
                  isChildContent: false,
                  destinationFolder: this.storageService.getStorageDestinationDirectoryPath(),
                  sourceFilePath: filePath,
                  correlationData: []
                }).pipe(
                  map((response: ContentImportResponse[]) => {
                    if (!response.length) {
                      this.commonUtilService.showToast('CONTENT_IMPORTED');
                      return;
                    }
                    response.forEach((contentImportResponse: ContentImportResponse) => {
                      switch (contentImportResponse.status) {
                        case ContentImportStatus.ALREADY_EXIST:
                          this.generateImportErrorTelemetry('CONTENT_ALREADY_EXIST');
                          this.commonUtilService.showToast('CONTENT_ALREADY_EXIST');
                          throw ContentImportStatus.ALREADY_EXIST;
                        case ContentImportStatus.IMPORT_FAILED:
                          this.generateImportErrorTelemetry('CONTENT_IMPORTED_FAILED');
                          this.commonUtilService.showToast('CONTENT_IMPORTED_FAILED');
                          throw ContentImportStatus.IMPORT_FAILED;
                        case ContentImportStatus.NOT_FOUND:
                          this.generateImportErrorTelemetry('NOT_FOUND');
                          this.commonUtilService.showToast('CONTENT_IMPORTED_FAILED');
                          throw ContentImportStatus.NOT_FOUND;
                      }
                    });
                  })
                )

**Comment by simranKtrigyn** [comment]
@swayangjit while investigating this issue, we observe that if the ecar file is downloded from the same device, import content functionality works. But if we receive ecar file externally (whatsapp, bluetooth share, drive share), IMPORT FAILED issue comes. I just want to know if this expected behaviour ??

Also FYI, the filepath we are getting while debugging is like this- file:///storage/emulated/0/Download/prshnottrii_1700289400353_do_3139280384301137921472_2 1.ecar
Check if this is the valid filepath

**Reply by chandu4002** [reply]
Hi @swayangjit , while investigating further in **sunbid-mobile-sdk** code, we are getting following issue
**{"success":false,"message":"decompress Operation fail"}**

Inside **importEcar** method, pls check following code for your reference
 
          return new ExtractEcar(this.fileService, this.zipService).execute(importContentContext);
          
          this.zipService.unzip(importContext.ecarFilePath, {target: directoryEntry.nativeURL}, () => {
                    resolve();
                }, (e) => {
                    console.log('error inside execute --1'+JSON.stringify(e));
                    reject(e);
                });

**Reply by swayangjit** [reply]
@chandu4002 Thanks for the detailed investigation. I will look into the issue and get back to you.

**Reply by swayangjit** [reply]
@chandu4002 The fix will be available in the upcoming releases and I will update this thread once the fix is available.

**Reply by chandu4002** [reply]
Hi @swayangjit Any update on this fix, can you please let me know ?
